### PR TITLE
Feature/addmanagecommand

### PR DESCRIPTION
**Pull request by pdubb29**
This adds the ability to manage dependencies.

### CODE REVIEW

**Comment by pdubb29** [comment]
this takes care of issue #148

File: GitDepend.UnitTests/Commands/AddCommandTests.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                Branch = "master",
+                DependencyDirectory = "..\\lib3",
+                Url = "https://test-url.com/repos/lib3"
+            };
+
+            var command = new AddCommand(options);
+
+            var code = command.Execute();
+
+            Assert.AreEqual(ReturnCode.Success, code);
```
Comment (kjjuno) [inline GitDepend.UnitTests/Commands/AddCommandTests.cs]: It looks like you are only ever testing the return code. There should also be a test to see that the config file was actually modified (or not modified) in the expected way.

File: GitDepend.UnitTests/Commands/RemoveCommandTests.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        [Test]
+        public void Execute_ShouldSucceed()
+        {
+            var algorithm = Container.Resolve<IDependencyVisitorAlgorithm>();
+            algorithm.Arrange(x => x.TraverseDependencies(new RemoveDependencyVisitor("Lib1"), Lib2Directory));
+        }
+
+        [Test]
+        public void Execute_ShouldFail_With_MisnamedDependency()
+        {
```
Comment (kjjuno) [inline GitDepend.UnitTests/Commands/RemoveCommandTests.cs]: This test needs to be finished, or deleted.

File: GitDepend.UnitTests/Visitors/RemoveDependencyVisitorTests.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        [SetUp]
+        public void TestSetup()
+        {
+            _visitor = new RemoveDependencyVisitor("lib1");
+        }
+
+        [Test]
+        public void RemoveDependencyShouldFindDependency()
+        {
+            var result = _visitor.VisitDependency("lib1", new Dependency()
```
Comment (kjjuno) [inline GitDepend.UnitTests/Visitors/RemoveDependencyVisitorTests.cs]: this looks like it's not correct. The first argument is the directory of the project.

This should probably be
```
_visitor.VisitDependency(Lib2Directory, Lib1Dependency)
```

File: GitDepend.UnitTests/Visitors/RemoveDependencyVisitorTests.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                            {
+                                Name = "lib1"
+                            }
+                        }
+                    },
+                    Name = "lib2",
+                }
+            });
+
+            Assert.AreEqual(ReturnCode.Success, result);
```
Comment (kjjuno) [inline GitDepend.UnitTests/Visitors/RemoveDependencyVisitorTests.cs]: Again, checking the return code is nice, but doesn't tell you anything about what the visitor did. To be truly helpful you need to do some other check to make sure that the visitor does the correct thing.

File: GitDepend.UnitTests/Visitors/RemoveDependencyVisitorTests.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+        [Test]
+        public void RemoveDependencyShouldReturnNotFound()
```
Comment (kjjuno) [inline GitDepend.UnitTests/Visitors/RemoveDependencyVisitorTests.cs]: This needs to be finished or deleted.

File: GitDepend/Commands/ManageCommand.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+        }
+
+        /// <summary>
+        /// Executes the command.
+        /// </summary>
+        /// <exception cref="System.NotImplementedException"></exception>
+        public ReturnCode Execute()
+        {
+            throw new NotImplementedException();
```
Comment (kjjuno) [inline GitDepend/Commands/ManageCommand.cs]: This command needs to be finished, or excluded

File: GitDepend/Commands/RemoveCommand.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        /// <exception cref="System.NotImplementedException"></exception>
+        public ReturnCode Execute()
+        {
+            var visitor = new RemoveDependencyVisitor(_options.DependencyName);
+
+            _algorithm.TraverseDependencies(visitor, _options.Directory);
+
+            if (visitor.ReturnCode == ReturnCode.NameDidNotMatchRequestedDependency)
+            {
+                Console.WriteLine(strings.ResourceManager.GetString("RET_NAME_DID_NOT_MATCH", CultureInfo.CurrentCulture));
```
Comment (kjjuno) [inline GitDepend/Commands/RemoveCommand.cs]: this should just be `strings.RET_NAME_DID_NOT_MATCH`

File: GitDepend/Visitors/IVisitor.cs [hunk 8]
```diff
@@ -22,7 +22,7 @@ public interface IVisitor
         ReturnCode VisitDependency(string directory, Dependency dependency);
 
         /// <summary>
-        /// Visists a project.
+        /// Visits a project.
```
Comment (kjjuno) [inline GitDepend/Visitors/IVisitor.cs]: nice catch! This is unfortunately probably in all the visitors.

File: GitDepend/Visitors/RemoveDependencyVisitor.cs [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        /// </summary>
+        /// <param name="directory">The directory of the project.</param>
+        /// <param name="config">The <see cref="GitDependFile" /> with project configuration information.</param>
+        /// <returns>
+        /// The return code.
+        /// </returns>
+        /// <exception cref="System.NotImplementedException"></exception>
+        public ReturnCode VisitProject(string directory, GitDependFile config)
+        {
+            if (_foundDependencyDirectory == string.Empty)
```
Comment (kjjuno) [inline GitDepend/Visitors/RemoveDependencyVisitor.cs]: please use `string.IsNullOrEmpty(...)`

File: GitDepend/Visitors/RemoveDependencyVisitor.cs [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+            string dir;
+            ReturnCode returnCode;
+
+            //visit dependency and get the configuration
+            var configFile = _factory.LoadFromDirectory(directory, out dir, out returnCode);
+            if (returnCode == ReturnCode.Success)
+            {
+                if (configFile.Name == _dependencyNameToRemove)
+                {
+                    _foundDependencyDirectory = dir;
```
Comment (kjjuno) [inline GitDepend/Visitors/RemoveDependencyVisitor.cs]: This seems wrong. `dir` is the directory where the projects `GitDepend.json` file was located. NOT the directory where the dependencies `GitDepend.json` file is. If you are trying to load the config for the dependency you need to use the given `directory` as the base, and the directory specified in the `Dependency` to calculate the dependency directory.

**Comment by pdubb29** [comment]
I'm adding tests for each class I've added. This is going to take me a while.

**Pull request merged**